proposals/multiple_user_script_worlds.md

```diff
@@ -0,0 +1,293 @@
+`worldId`, to the `WorldProperties` and `RegisteredUserScript` types, as below.
+
+Relevant methods and types:
+
+```diff
+ export namespace userScripts {
+   export interface WorldProperties {
++    /**
++     * Specifies the ID of the specific user script world to update.
++     * If not provided, updates the properties of the default user script
++     * world.
++     * Values with leading underscores (`_`) are reserved.
++     */
++    worldId?: string;
+
+     /**
+      * Specifies the world's CSP. The default is the `ISOLATED` world CSP.
+      */
+     csp?: string;
+
+     /**
+      * Specifies whether messaging APIs are exposed. The default is `false`.
+      */
+     messaging?: boolean;
+   }
+
+   /**
+    * Properties for a registered user script.
+    */
+   export interface RegisteredUserScript {
+     /**
+      * If true, it will inject into all frames, even if the frame is not the
+      * top-most frame in the tab. Each frame is checked independently for URL
+      * requirements; it will not inject into child frames if the URL
+      * requirements are not met. Defaults to false, meaning that only the top
+      * frame is matched.
+      */
+     allFrames?: boolean;
+
+     /**
+      * Specifies wildcard patterns for pages this user script will NOT be
+      * injected into.
+      */
+     excludeGlobs?: string[];
+
+     /**
+      * Excludes pages that this user script would otherwise be injected into.
+      * See Match Patterns for more details on the syntax of these strings.
+      */
+     excludeMatches?: string[];
+
+     /**
+      * The ID of the user script specified in the API call. This property
+      * must not start with a '_' as it's reserved as a prefix for generated
+      * script IDs.
+      */
+     id: string;
+
+     /**
+      * Specifies wildcard patterns for pages this user script will be
+      * injected into.
+      */
+     includeGlobs?: string[];
+
+     /**
+      * The list of ScriptSource objects defining sources of scripts to be
+      * injected into matching pages.
+      */
+     js: ScriptSource[];
+
+     /**
+      * Specifies which pages this user script will be injected into. See
+      * Match Patterns for more details on the syntax of these strings. This
+      * property must be specified for ${ref:register}.
+      */
+     matches?: string[];
+
+     /**
+      * Specifies when JavaScript files are injected into the web page. The
+      * preferred and default value is document_idle
+      */
+     runAt?: RunAt;
+
+     /**
+      * The JavaScript execution environment to run the script in. The default
+      * is `USER_SCRIPT`
+      */
+     world?: ExecutionWorld;
+
++    /**
++     * If specified, specifies a specific user script world ID to execute in.
++     * Only valid if `world` is omitted or is `USER_SCRIPT`. If omitted, the
++     * script will execute in the default user script world.
++     * Values with leading underscores (`_`) are reserved.
```

Comment:
optional nit: Maybe require **two** leading underscores?  It seems that double-leading-underscore is less likely to be used in practice than a single leading underscore.  And this would also be [consistent with C++](https://wiki.sei.cmu.edu/confluence/display/cplusplus/DCL51-CPP.+Do+not+declare+or+define+a+reserved+identifier#:~:text=Each%20name%20that%20contains%20a,name%20in%20the%20global%20namespace.).

QUESTION: Just to double-check my understanding - reserved world ids may lead to some undesirable behavior **within** an extension, but this is okay, because they won't have any effect **outside** of an extension.  In other words, a malicious extension won't be able to abuse reserved world ids to somehow cross a security boundary somewhere, right?

Reply:
Thanks, Lukasz!

> optional nit: Maybe require two leading underscores?

Unfortunately, we have a decent amount of precedent in extension APIs for using single-underscore for reserved (a few different APIs, file paths in the extension directory, etc).  We might revisit that at some point in the future, but I think being consistent is best here for now.

> In other words, a malicious extension won't be able to abuse reserved world ids to somehow cross a security boundary somewhere, right?

They certainly shouldn't be able to! ; )  I'd expect any implementation to simply reject any call from an extension that specifies a world ID beginning with '_'.  The reservation of '_' IDs was just a suggestion in case implementors wanted to allow all user script worlds to have a world ID (e.g., using "_default" for the default world in the underlying implementation).  On the Chromium side, we just use a std::optional<>, so there's really no way an extension should be able to abuse this.